pyintegration/tests/test_dtypes.py

```diff
@@ -0,0 +1,194 @@
+        self.assertEqual(DTypes.float_, DTypes.single)
+        self.assertEqual(DTypes.float_, DTypes.float32)
+        self.assertEqual(DTypes.double, DTypes.float64)
+
+    def test_j_type(self):
+        self.assertEqual(dtypes.bool_.j_type, jpy.get_type("java.lang.Boolean"))
+        self.assertEqual(dtypes.byte.j_type, jpy.get_type("byte"))
+        self.assertEqual(dtypes.short.j_type, jpy.get_type("short"))
+        self.assertEqual(dtypes.char.j_type, jpy.get_type("char"))
+        self.assertEqual(dtypes.int_.j_type, jpy.get_type("int"))
+        self.assertEqual(dtypes.long.j_type, jpy.get_type("long"))
+        self.assertEqual(dtypes.float_.j_type, jpy.get_type("float"))
+        self.assertEqual(dtypes.double.j_type, jpy.get_type("double"))
+        self.assertEqual(dtypes.string.j_type, jpy.get_type("java.lang.String"))
+        self.assertEqual(dtypes.BigDecimal.j_type, jpy.get_type("java.math.BigDecimal"))
+
+    def test_qst_type(self):
+        self.assertIn(".qst.type.", str(dtypes.short.qst_type))
+
+    def test_custom_type(self):
+        self.assertIn("CustomType", str(dtypes.StringSet.qst_type))
+
+    def test_period(self):
+        hour_period = dtypes.Period.j_type("T1H")
+        self.assertTrue(isinstance(hour_period, dtypes.Period.j_type))
+
+    def test_callable(self):
+        big_decimal = dtypes.BigDecimal(12.88)
+        self.assertIn("12.88", str(big_decimal))
+
+        big_decimal2 = dtypes.BigDecimal("12.88")
+        self.assertIn("12.88", str(big_decimal2))
+
+    def test_array(self):
+        j_array = dtypes.int_.array(5)
+        for i in range(5):
+            j_array[i] = i
+        np_array = numpy.frombuffer(j_array, numpy.int32)
+        self.assertTrue((np_array == numpy.array([0, 1, 2, 3, 4], dtype=numpy.int32)).all())
+
+    def test_array_of(self):
+        j_array = dtypes.int_.array_of(range(5))
+        np_array = numpy.frombuffer(j_array, numpy.int32)
+        self.assertTrue((np_array == numpy.array([0, 1, 2, 3, 4], dtype=numpy.int32)).all())
+
+        j_array = dtypes.int_.array_of([0, 1, 2, 3, 4])
+        np_array = numpy.frombuffer(j_array, numpy.int32)
+        self.assertTrue((np_array == numpy.array([0, 1, 2, 3, 4], dtype=numpy.int32)).all())
+
+    def test_integer_array_of(self):
+        np_array = np.array([float('nan'), NULL_DOUBLE, 1.123, np.inf], dtype=np.float64)
+
+        with self.subTest("long array from numpy array/pd series with nan_to_value"):
+            expected = [0, NULL_LONG, 1, NULL_LONG]
+            j_array = dtypes.int64.array_of(np_array, nan_to_value=0)
+            self.assertIn("[J", str(type(j_array)))
+            py_array = [x for x in j_array]
+            self.assertEqual(expected, py_array)
+
+            pd_series = pd.Series(np_array)
+            j_array = dtypes.int64.array_of(pd_series)
+            self.assertIn("[J", str(type(j_array)))
+            expected = [NULL_LONG, NULL_LONG, 1, NULL_LONG]
```

Comment:
This looks like it is handling the `NULL_DOUBLE`->`NULL_LONG` conversion properly, which I was concerned about.  It does bring up another problem.  We're mapping `Inf`->`NULL`, which may be problematic and may need handling.

Same comment in other places below.